### PR TITLE
Product Bar & Product Pane

### PR DESCRIPTION
**Pull request by vesln**
product-bar

## Scope

- Restructure as micro apps
- Implement product bar with social buttons
- Implement product pane, currently using `/posts/xyz`

## Notes

- Hit a bunch of iframe related bugs in react... I'm going to spend more time investigating, instead of implementing hacks when time permits
- I'd like to extract `ProductHunt` as a standalone browser client in the near future
- It'd be great if `discussion_url` returns the url with `https` scheme
- The caching of access tokens is not extremely useful, since the `content` app runs in the context of the current website
- Test coverage can be improved, but that will happen in next iterations
- UX is the simplest I could've implemented. Feel free to give it a try and make suggestions

### CODE REVIEW

File: src/apps/content/actions/product-actions.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+/**
+ * Dependencies.
```
Comment (andreasklinger) [inline src/apps/content/actions/product-actions.js]: In the web app we use the top lines to explain in 1 sentence what the component is supposed to do

in most cases it is obvious but still helps a bit to get into the context of the application

eg

```
# This component lazy loads images (duh…)
#
# Consists of two components
#   - The container that does the actual lazyloading
#   - The elements send events to notify the container of their existence
#
#
# Usage:
#   <body data-component="LazyLoadContainer">
#     <img data-component="LazyLoadElement" data-lazyload-url="/images/foobar.png"/>
#   </body>
#
```

or
```
# Component displaying prerendered content into a fullscreen overlay
#
# Usage:
#
#   <div data-component="FullscreenOverlayContent">
#      ...
#   </div>
#
```

File: src/apps/content/actions/product-actions.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+let AppDispatcher = require('../dispatcher/app-dispatcher');
```
Comment (andreasklinger) [inline src/apps/content/actions/product-actions.js]: lowlevel stuff:

we are now using `-` `_` and `camelCase` for js
which one is the "right" one ? 

cc @RStankov @korbin

File: src/apps/content/actions/product-actions.js (hunk 1 above)
Comment (andreasklinger) [inline src/apps/content/actions/product-actions.js]: ahh you do it sometimes but next to the class

```diff
+ let BodyModifier = React.createClass({
+
+  /**
+   * When we mount the component we add the custom
+   * body class.
+   *
+   * @public
+   */
```

didnt see it there

File: src/apps/content/components/product-bar.react.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        <BodyModifier className={BODY_CLASS} />
+        <TopElements />
+        <Pane bodyClass={OVERLAY_BODY_CLASS} url={url} onClick={this._togglePane} />
+
+        <Frame className="__phc-bar" id={PRODUCT_BAR_ID} scrolling="no" head={
+          <link type='text/css' rel='stylesheet' href={CSS_URL} />
+        }>
+          <ProductDetails product={this.state.product} onClick={this._togglePane} />
+
+          <div className="facebook">
```
Comment (andreasklinger) [inline src/apps/content/components/product-bar.react.js]: just to ensure - we do not run into some css leaking problems here - right?

File: src/apps/content/util/get-host.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ * @param {String} url
+ * @returns {String}
+ * @public
+ */
+
+function getHost(url) {
+  let parser = document.createElement('a');
+  parser.href = url;
+  return parser.host;
```
Comment (andreasklinger) [inline src/apps/content/util/get-host.js]: TIL

**Comment by andreasklinger** [comment]
Added the `ready for review` label as i assume you just forgot it

:+1: from me 
like the structure a lot

any way we can test it in the browsers easily? shall i download it or do you want to do a google hosted dev version anyhow?

File: src/apps/content/actions/product-actions.js (hunk 2 above)
Comment (RStankov) [inline src/apps/content/actions/product-actions.js]: I thought we were using `_` in the web app.

Comment (vesln) [inline src/apps/content/actions/product-actions.js]: React's convention is `TitleCase`, the most common node.js one is with `-` and `_` is also an option. I personally don't care much, so lmk what we should continue with.

File: src/apps/content/actions/product-actions.js (hunk 1 above)
Comment (vesln) [inline src/apps/content/actions/product-actions.js]: Point taken, will be improved

File: src/apps/content/components/product-bar.react.js (hunk 3 above)
Comment (vesln) [inline src/apps/content/components/product-bar.react.js]: It's inside an iframe. The css we are exposing to clients is prefixed with `__phc`.

**Comment by vesln** [comment]
We should setup a "nightly" build at some point, but I don't think it's a huge priority atm. Let's think about it and do it properly once we have all features for 0.2 in.

File: src/apps/content/actions/product-actions.js (hunk 2 above)
Comment (korbin) [inline src/apps/content/actions/product-actions.js]: TitleCase (standard node) for non-rails stuff. Rails-API responses - camelCase.

**Comment by vesln** [comment]
@andreasklinger I've address the React component documentation, by using the following format:

Name
Description
Usage
Properties (if any)
State (if any)

Regarding file naming, will keep it as it is until a final decision has been made -- shouldn't be a blocker for this pr.

**Pull request merged**